### ISSUE TITLE
 Connected to the stun server by TCP to receive the TCP RST message

### ISSUE THREAD

**Issue by luoyefeisong** [coturn/coturn#349]
**turnserver address:** 132.232.3.109:12345
**start command:** turnserver -S -V

**capture:**
![image](https://user-images.githubusercontent.com/13747710/53480802-e3187300-3ab6-11e9-966e-e429d26305c7.png)

**log of turnserver:**
26: timer_event_handler: timeout 0x7f24a4001c90: timer_handler
26: timer_event_handler: timeout 0x7f24a4002df0: timer_timeout_handler
26: timer_event_handler: timeout 0x7f24ac001c90: timer_handler
26: timer_event_handler: timeout 0x7f24ac002df0: timer_timeout_handler
26: timer_event_handler: timeout 0x7f2494001e30: timer_handler
26: server_input_handler:74:start
26: IPv4. tcp or tls connected to: 182.150.22.249:2399
26: server_input_handler:121:end
26: open_client_connection_session:4601:start
26: open_client_connection_session:4635:end
26: read_client_connection:4402:start
26: read_client_connection: data.buffer=0x7f24a40088dc, data.len=36
26: handle_turn_binding:2693:start
26: handle_turn_binding:2880:end
26: session 001000000000000001: realm <> user <>: incoming packet BINDING processed, success
26: write_client_connection:4188:start
26: write_client_connection: prepare to write to s 0x7f24a4003450
26: write_client_connection:4211:end
26: read_client_connection:4508:end
26: timer_event_handler: timeout 0x5636de977fc0: timer_handler
27: timer_event_handler: timeout 0x5636de9781b0: rtcp_map_timeout_handler
27: timer_event_handler: timeout 0x7f24ac001c90: timer_handler
27: timer_event_handler: timeout 0x7f24ac002df0: timer_timeout_handler
27: timer_event_handler: timeout 0x7f24a4001c90: timer_handler
27: timer_event_handler: timeout 0x7f24a4002df0: timer_timeout_handler
27: timer_event_handler: timeout 0x7f2494001e30: timer_handler
27: timer_event_handler: timeout 0x5636de977fc0: timer_handler
28: timer_event_handler: timeout 0x7f24ac001c90: timer_handler
28: timer_event_handler: timeout 0x7f24ac002df0: timer_timeout_handler
28: timer_event_handler: timeout 0x7f24a4001c90: timer_handler
28: timer_event_handler: timeout 0x7f24a4002df0: timer_timeout_handler
28: timer_event_handler: timeout 0x7f2494001e30: timer_handler
28: timer_event_handler: timeout 0x5636de977fc0: timer_handler
29: timer_event_handler: timeout 0x7f24ac001c90: timer_handler
29: timer_event_handler: timeout 0x7f24ac002df0: timer_timeout_handler
29: timer_event_handler: timeout 0x7f24a4001c90: timer_handler
29: timer_event_handler: timeout 0x7f24a4002df0: timer_timeout_handler
29: timer_event_handler: timeout 0x7f2494001e30: timer_handler
29: timer_event_handler: timeout 0x7f24a40083d0: client_to_be_allocated_timeout_handler
29: client_to_be_allocated_timeout_handler:4153:start
29: shutdown_client_connection:4068:start
29: closing session 0x7f24a40035c0, client socket 0x7f24a4003450 (socket session=0x7f24a40035c0)
29: session 001000000000000001: closed (2nd stage), user <> realm <> origin <>, local 172.27.0.4:12345, remote 182.150.22.249:2399, reason: allocation watchdog determined stale session state
29: shutdown_client_connection:4133:end
29: client_to_be_allocated_timeout_handler:4183:end

**Comment by misi** [comment]
Can you please point out your issue? I don't see any problem with this binding request.  Your client did open a tcp session and after transmission you didn't close the session, so the server closed it after a timeout.

**Comment by luoyefeisong** [comment]
@misi 
My problem is that the client sends a binding request every 15 seconds, but it still times out after 60 seconds.
I change the start command "turnserver -S -V" to “turnserver -V”，because I read the code found that it will time out after two seconds after setting -S.
My capture:
![image](https://user-images.githubusercontent.com/13747710/53482721-db5acd80-3aba-11e9-994c-a5e8087a3100.png)

**Comment by luoyefeisong** [comment]
@misi 
I debugged, the timeout function [client_to_be_allocated_timeout_handler] closes the connection at this location:
![image](https://user-images.githubusercontent.com/13747710/53559579-5c7e9700-3b85-11e9-8f95-fbc4408b36bd.png)
rs4 is null. the content of ss->alloc should be the allocation process of the turn, but I am sending a binding request.
I think we should set a separate timeout handler for the binding request in the TCP connection.

**Comment by misi** [comment]
60 and 3 sec is the timeout:
https://github.com/coturn/coturn/blob/0a1ec372f6f18891c4ee6d36c98a8cc8b730afd9/src/server/ns_turn_server.c#L76-L77

STUN Binding is a request response protocol.
I guess it doesn't make a session. I am not sure if it should make..

**Comment by luoyefeisong** [comment]
@misi 
https://github.com/coturn/coturn/blob/0a1ec372f6f18891c4ee6d36c98a8cc8b730afd9/src/server/ns_turn_server.c#L4203

It only has a relay session, don't have stun session.

**Comment by misi** [comment]
UDP would be much much more ideal for STUN (binding request/response).
Is TCP mandatory for you? 

Oleg wrote STUN over TCP without session. It should be a reason why it is implemented this way.
My guess to create for TCP a session is an overhead that needs more cpu/mem and may a way to exhaust the server resources. 

Anyhow we should investigate it further.

Many thanks for your cooperation!!

**Comment by misi** [comment]
I agree that coturn should support multiple TCP STUN transaction on a single  tcp connection.
Can you send a pull request about it. 
(Unfortunately I am busy with other fixes..)

rfc5389 allows it:
```
The client MAY send multiple transactions over a single TCP (or TLS-
   over-TCP) connection, and it MAY send another request before
   receiving a response to the previous. 
```

**Comment by luoyefeisong** [comment]
I really need TCP, so I don't consider the message ordering. 
I am still not very familiar with the code, I will send you a pull request later.

Currently, I am simply handling：
1. Add a flag, set the flag to TRUE when receiving the binding request.
2. When entering the timeout function, if the flag is TRUE, do not close the client session, just set the flag to FALSE.
![image](https://user-images.githubusercontent.com/13747710/53564956-2811d780-3b93-11e9-9482-c7682c34c75a.png)

**Comment by luoyefeisong** [comment]
Also, thank you for taking the time to reply.
